### PR TITLE
Blame for vstest

### PR DESCRIPTION
**Pull request by vagisha-nidhi**
This feature gets the faulty test case name in event of a test host crash of reason unknown.
There are various scenarios in which the test host crashes due to a faulty test case and the reason cannot be diagnosed.
This pull request gets the name of the test case which caused the crash and displays it on the console.
To enable the blame mode, use --Blame|/Blame

The usage is as follows -

'''
"vstest.console.exe" /Testadapterpath:"C:\TestProject\TestProject\bin\Debug\net46"  C:\path\to\your\assembly.dll  /Blame
'''
crsh

### CODE REVIEW

File: src/Microsoft.TestPlatform.BlameDataCollector/BlameLogger.cs [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+            : this(ConsoleOutput.Instance, new XmlReaderWriter())
+        {
+        }
+
+        /// <summary>
+        /// Constructor added for testing purpose
+        /// </summary>
+        /// <param name="output"></param>
+        /// <param name="blameReaderWriter">BlameReaderWriter Instance</param>
+        public BlameLogger(IOutput output, IBlameReaderWriter blameReaderWriter)
```
Comment (codito) [inline src/Microsoft.TestPlatform.BlameDataCollector/BlameLogger.cs]: Should this be public?

File: src/Microsoft.TestPlatform.BlameDataCollector/BlameLogger.cs [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+        {
+            this.Output = output;
+            this.blameReaderWriter = blameReaderWriter;
+        }
+
+        protected IOutput Output
```
Comment (codito) [inline src/Microsoft.TestPlatform.BlameDataCollector/BlameLogger.cs]: Why protected?

File: src/Microsoft.TestPlatform.BlameDataCollector/BlameLogger.cs [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+
+        /// <summary>
+        /// Called when a test run is complete.
+        /// </summary>
+        private void TestRunCompleteHandler(object sender, TestRunCompleteEventArgs e)
+        {
+            ValidateArg.NotNull<object>(sender, "sender");
+            ValidateArg.NotNull<TestRunCompleteEventArgs>(e, "e");
+
+            if (!e.IsAborted) return;
```
Comment (codito) [inline src/Microsoft.TestPlatform.BlameDataCollector/BlameLogger.cs]: nit: please wrap if clauses in parenthesis.

File: src/Microsoft.TestPlatform.BlameDataCollector/BlameLogger.cs [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+        /// Fetches faulty test case
+        /// </summary>
+        private string GetFaultyTestCase(TestRunCompleteEventArgs e)
+        {
+            foreach (var attachmentSet in e.AttachmentSets)
+            {
+                if (attachmentSet.DisplayName.Equals(Constants.BlameDataCollectorName))
+                {
+                    var filepath = attachmentSet.Attachments[0].Uri.LocalPath;
+                    List<TestCase> testCaseList = this.blameReaderWriter.ReadTestSequence(filepath);
```
Comment (codito) [inline src/Microsoft.TestPlatform.BlameDataCollector/BlameLogger.cs]: Can we use `Enumerable.LastOrDefault`?

File: src/Microsoft.TestPlatform.BlameDataCollector/Constants.cs [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+        
+        /// <summary>
+        /// Friendly name of the data collector
+        /// </summary>
+        public const string BlameDataCollectorName = "Blame";
+
+        /// <summary>
+        /// Test run Abort 
+        /// </summary>
+        public const string TestRunAbortReason = "The test run was aborted because the host process crashed unexpectedly while executing test ";
```
Comment (codito) [inline src/Microsoft.TestPlatform.BlameDataCollector/Constants.cs]: Not used?

File: src/Microsoft.TestPlatform.BlameDataCollector/Friends.cs [hunk 6]
```diff
@@ -0,0 +1,14 @@
...
+
+using System.Runtime.CompilerServices;
+
+#region Product Assemblies
+
+#endregion
+
+#region Test Assemblies
+
+[assembly: InternalsVisibleTo("Microsoft.TestPlatform.BlameDataCollector.UnitTests, PublicKey=002400000480000094000000060200000024000052534131000400000100010007d1fa57c4aed9f0a32e84aa0faefd0de9e8fd6aec8f87fb03766c834c99921eb23be79ad9d5dcc1dd9ad236132102900b723cf980957fc4e177108fc607774f29e8320e92ea05ece4e821c0a5efe8f1645c4c0c93c1ab99285d622caa652c1dfad63d745d6f2de5f17e5eaf0fc4963d261c8a12436518206dc093344d5ad293")]
```
Comment (codito) [inline src/Microsoft.TestPlatform.BlameDataCollector/Friends.cs]: Not sure if this required.

File: test/Microsoft.TestPlatform.BlameDataCollector.UnitTests/BlameCollectorTests.cs [hunk 7]
```diff
@@ -0,0 +1,129 @@
...
+            Assert.ThrowsException<ArgumentNullException>(() =>
+            {
+                this.blameDataCollector.Initialize(this.configurationElement,
+                    this.mockDataColectionEvents.Object, this.mockDataCollectionSink.Object,
+                    (DataCollectionLogger)null, null);
+            });
+        }
+
+        [TestMethod]
+        public void TriggerTestCaseStartedHandlerShouldIncreaseTestStartCount()
```
Comment (codito) [inline test/Microsoft.TestPlatform.BlameDataCollector.UnitTests/BlameCollectorTests.cs]: This test is testing an internal implementation detail. We should care about the user behavior.

**Review by codito — CHANGES_REQUESTED** [review]
Suggest a few changes.

File: src/Microsoft.TestPlatform.Extensions.BlameDataCollector/BlameCollector.cs [hunk 8]
```diff
@@ -0,0 +1,145 @@
...
+            if (EqtTrace.IsInfoEnabled)
+            {
+                EqtTrace.Info("Blame Collector : Session End");
+            }
+
+            // If the last test crashes, it will not invoke a test case end and therefore 
+            // In case of crash testStartCount will be greater than testEndCount and we need to send write the sequence
+            if (this.testStartCount > this.testEndCount)
+            {
+                var filepath = Path.Combine(AppContext.BaseDirectory, Constants.AttachmentFileName);
```
Comment (vagisha-nidhi) [inline src/Microsoft.TestPlatform.Extensions.BlameDataCollector/BlameCollector.cs]: Check filepath

File: src/Microsoft.TestPlatform.Extensions.BlameDataCollector/BlameLogger.cs [hunk 9]
```diff
@@ -0,0 +1,149 @@
...
+    [FriendlyName(BlameLogger.FriendlyName)]
+    [ExtensionUri(BlameLogger.ExtensionUri)]
+    public class BlameLogger : ITestLogger
+    {
+        #region Constants
+
+        /// <summary>
+        /// Uri used to uniquely identify the Blame logger.
+        /// </summary>
+        public const string ExtensionUri = "logger://Microsoft/TestPlatform/Extensions/Blame";
```
Comment (vagisha-nidhi) [inline src/Microsoft.TestPlatform.Extensions.BlameDataCollector/BlameLogger.cs]: Version

File: src/Microsoft.TestPlatform.Extensions.BlameDataCollector/BlameLogger.cs [hunk 10]
```diff
@@ -0,0 +1,149 @@
...
+                {
+                    var uriDataAttachment = attachmentSet.Attachments.LastOrDefault();
+                    if (uriDataAttachment != null)
+                    {
+                        var filepath = uriDataAttachment.Uri.LocalPath;
+                        var testCaseList = this.blameReaderWriter.ReadTestSequence(filepath);
+                        if (testCaseList.Count > 0)
+                        {
+                            var testcase = testCaseList[testCaseList.Count - 1];
+                            return testcase.FullyQualifiedName;
```
Comment (vagisha-nidhi) [inline src/Microsoft.TestPlatform.Extensions.BlameDataCollector/BlameLogger.cs]: LastOrDefault

File: src/Microsoft.TestPlatform.Extensions.BlameDataCollector/Microsoft.TestPlatform.Extensions.BlameDataCollector.csproj [hunk 11]
```diff
@@ -0,0 +1,53 @@
...
+    <AssemblyName>Microsoft.TestPlatform.Extensions.BlameDataCollector</AssemblyName>
+    <TargetFrameworks>net46;netstandard1.5</TargetFrameworks>
+  </PropertyGroup>
+  <ItemGroup>
+    <EmbeddedResource Include="Resources\Resources.resx" />
+  </ItemGroup>
+  <ItemGroup>
+    <ProjectReference Include="..\Microsoft.TestPlatform.ObjectModel\Microsoft.TestPlatform.ObjectModel.csproj" />
+    <ProjectReference Include="..\..\src\Microsoft.TestPlatform.CoreUtilities\Microsoft.TestPlatform.CoreUtilities.csproj" />
+  </ItemGroup>
```
Comment (vagisha-nidhi) [inline src/Microsoft.TestPlatform.Extensions.BlameDataCollector/Microsoft.TestPlatform.Extensions.BlameDataCollector.csproj]: Enable stylecop

File: src/package/sign/sign.proj [hunk 12]
```diff
@@ -129,6 +129,7 @@
       <!-- NetCoreExtensions -->
       <CoreAssembliesToSign Include="$(ArtifactsCoreDirectory)Extensions\Microsoft.VisualStudio.TestPlatform.Extensions.Trx.TestLogger.dll" />
 	  <CoreAssembliesToSign Include="$(ArtifactsCoreDirectory)Extensions\Microsoft.TestPlatform.TestHostRuntimeProvider.dll" />
+	  <CoreAssembliesToSign Include="$(ArtifactsCoreDirectory)Extensions\Microsoft.TestPlatform.Extensions.BlameDataCollector.dll" />
```
Comment (vagisha-nidhi) [inline src/package/sign/sign.proj]: Add for net46

**Review by codito — APPROVED** [review]

**Pull request merged**